presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -83,6 +117,18 @@ public boolean equals(Object obj)
     @Override
     public String toString()
     {
-        return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(timeZoneKey.getId())).format(formatter);
+        // FIXME This is hack that we need to use as the timezone interpretation depends on date (not only on time)
+        // Additionally we cannot just grab current TZ offset from timezoneOffsetReferenceTimestampUtc and use it
+        // as our formatting library will fail to display expected TZ symbol.
+        long currentMillisOfDay =
+                LocalTime.from(
+                        Instant.ofEpochMilli(timezoneOffsetReferenceTimestampUtc)
+                               .atZone(ZoneId.of(TimeZoneKey.UTC_KEY.getId())))
+                        .toNanoOfDay() / 1_000_000L; // nanos to millis
+        long timeMillisUtcInCurrentDay = timezoneOffsetReferenceTimestampUtc - currentMillisOfDay + millisUtc;
+
+        Instant utcInstantInCurrentDay = Instant.ofEpochMilli(timeMillisUtcInCurrentDay);
+        ZonedDateTime zonedDateTime = utcInstantInCurrentDay.atZone(ZoneId.of(timeZoneKey.getId()));
+        return zonedDateTime.format(formatter);
```

Comment:
Lots and lots of test, sadly mostly indirect :(
I think we could add some testing to that, especially checking if this is aligned with TIME WITH TIME ZONE to VARCHAR cast, which is expected but not guaranteed by the code.

I'll add that to TODO for later or escalating to further tickets - depending on circumstances.